packages-exp/auth-exp/src/core/strategies/credential.test.ts

```diff
@@ -63,32 +69,115 @@ describe('core/strategies/signInWithCredential', () => {
   );
 
   let auth: Auth;
+  let getAccountInfoEndpoint: mockFetch.Route;
+  let user: User;
 
   beforeEach(async () => {
     auth = await testAuth();
     mockFetch.setUp();
     authCredential._setIdTokenResponse(idTokenResponse);
-    mockEndpoint(Endpoint.GET_ACCOUNT_INFO, {
+    getAccountInfoEndpoint = mockEndpoint(Endpoint.GET_ACCOUNT_INFO, {
       users: [serverUser]
     });
+
+    user = testUser(auth, 'uid', undefined, true);
   });
+
   afterEach(mockFetch.tearDown);
 
-  it('should return a valid user credential', async () => {
-    const { credential, user, operationType } = await signInWithCredential(
-      auth,
-      authCredential
-    );
-    expect(credential!.providerId).to.eq(ProviderId.FIREBASE);
-    expect(credential!.signInMethod).to.eq(SignInMethod.EMAIL_LINK);
-    expect(user.uid).to.eq('local-id');
-    expect(user.tenantId).to.eq('tenant-id');
-    expect(user.displayName).to.eq('display-name');
-    expect(operationType).to.eq(OperationType.SIGN_IN);
+  describe('signInWithCredential', () => {
+    it('should return a valid user credential', async () => {
+      const { credential, user, operationType } = await signInWithCredential(
+        auth,
+        authCredential
+      );
+      expect(credential!.providerId).to.eq(ProviderId.FIREBASE);
+      expect(credential!.signInMethod).to.eq(SignInMethod.EMAIL_LINK);
+      expect(user.uid).to.eq('local-id');
+      expect(user.tenantId).to.eq('tenant-id');
+      expect(user.displayName).to.eq('display-name');
+      expect(operationType).to.eq(OperationType.SIGN_IN);
+    });
+
+    it('should update the current user', async () => {
+      const { user } = await signInWithCredential(auth, authCredential);
+      expect(auth.currentUser).to.eq(user);
+    });
+  });
+
+  describe('linkWithCredential', () => {
+    it('should throw an error if the provider is already linked', async () => {
+      getAccountInfoEndpoint.response = {
+        users: [
+          {
+            ...serverUser,
+            providerUserInfo: [{ providerId: ProviderId.FIREBASE }]
+          }
+        ]
+      };
+
+      await expect(linkWithCredential(user, authCredential)).to.be.rejectedWith(
+        FirebaseError,
+        'Firebase: User can only be linked to one identity for the given provider. (auth/provider-already-linked).'
+      );
+    });
+
+    it('should return a valid user credential', async () => {
+      const {
+        credential,
+        user: newUser,
+        operationType
+      } = await linkWithCredential(user, authCredential);
+      expect(operationType).to.eq(OperationType.LINK);
+      expect(newUser).to.eq(user);
+      expect(credential).to.be.null;
+    });
   });
 
-  it('should update the current user', async () => {
-    const { user } = await signInWithCredential(auth, authCredential);
-    expect(auth.currentUser).to.eq(user);
+  describe('_assertLinkedStatus', () => {
```

Comment:
since all the phone stuff behaves so differently, can you add a separate set of tests for linking a non-phone cred?

Reply:
Done. I just switched these over to GOOGLE.. The provider in question doesn't actually matter for this set of tests